src/DeviceInput/deviceInputSystem.ts

```diff
@@ -291,13 +307,42 @@ export class DeviceInputSystem implements IDisposable {
             if (kbKey) {
                 kbKey[evt.keyCode] = 0;
                 if (this.onInputChanged) {
-                    this.onInputChanged(DeviceType.Keyboard, 0, evt.keyCode, 1, kbKey[evt.keyCode]);
+                    this.onInputChanged(DeviceType.Keyboard, 0, evt.keyCode, 1, kbKey[evt.keyCode], evt);
+                }
+            }
+        });
+
+        let attachedFunction = () => {
+            this._elementToAttachTo.addEventListener("keydown", this._keyboardDownEvent);
+            this._elementToAttachTo.addEventListener("keyup", this._keyboardUpEvent);
+        };
+
+        this._onCanvasFocusObserver = this._engine.onCanvasFocusObservable.add(
+            (() => {
+                if (document.activeElement === this._elementToAttachTo) {
+                    attachedFunction();
+                }
+                return attachedFunction;
+            })()
+        );
+
+        this._onCanvasBlurObserver = this._engine.onCanvasBlurObservable.add(() => {
+            if (!this._elementToAttachTo) {
+                return;
+            }
+
+            this._elementToAttachTo.removeEventListener("keydown", this._keyboardDownEvent);
+            this._elementToAttachTo.removeEventListener("keyup", this._keyboardUpEvent);
+
+            // Reset all keyboard inputs
+            if (this._inputs[DeviceType.Keyboard][0]) {
+                for(let i = 0; i < DeviceInputSystem._MAX_KEYCODES; i++) {
+                    this._inputs[DeviceType.Keyboard][0][i] = 0;
                 }
             }

```

Comment:
Does this need to run if this.elementToAttachTo if falsy?

Reply:
No longer applicable